File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/entity/EntityActor.java
package net.bestia.zoneserver.actor.entity;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.PoisonPill;
import akka.actor.Terminated;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;
import net.bestia.messages.EntityMessage;
import net.bestia.messages.entity.ComponentEnvelope;
import net.bestia.messages.entity.EntityComponentStateMessage;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.util.Objects;

import static net.bestia.messages.entity.EntityComponentStateMessage.ComponentState.INSTALL;

/**
 * The {@link EntityActor} is a persistent actor managing all aspects of a
 * spawned entity. This means it will keep references to AI actors or attached
 * component actors. This actor has to be used as an persisted shared actor.
 *
 * If there are no more active component actors this actor will cease operation.
 * However incoming component messages will restart this actor and attach the
 * component actor once more.
 * 
 * This actor has to react on certain incoming request messages like for example
 * attaching a ticking script to the entity.
 * 
 * @author Thomas Felix
 *
 */
@Component
@Scope("prototype")
public class EntityActor extends AbstractActor {
	private final LoggingAdapter LOG = Logging.getLogger(getContext().getSystem(), this);

	private final EntityComponentActorFactory factory;

	/**
	 * Saves the references for the actors managing components.
	 */
	private BiMap<Long, ActorRef> componentActors = HashBiMap.create();

	private long entityId;

	@Autowired
	public EntityActor(EntityComponentActorFactory factory) {

		this.factory = Objects.requireNonNull(factory);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder()
				.match(EntityComponentStateMessage.class, this::handleComponentMessage)
				.match(ComponentEnvelope.class, this::handleComponentPayload)
				.match(Terminated.class, this::handleTerminated)
				.build();
	}

	/**
	 * Checks if we have such a associated component and if so delivers the
	 * message.
	 */
	private void handleComponentPayload(ComponentEnvelope msg) {
		checkStartup(msg);

		final long compId = msg.getComponentId();

		if (!componentActors.containsKey(compId)) {
			LOG.debug("Component message unhandled: {}.", msg);
			unhandled(msg);
			return;
		}

		final ActorRef compActor = componentActors.get(compId);
		LOG.debug("Forwarding comp message: {} to: {}.", msg, compActor);
		compActor.tell(msg.getPayload(), getSelf());
	}

	/**
	 * Adds or removes a active component actor from the entity.
	 */
	private void handleComponentMessage(EntityComponentStateMessage msg) {
		checkStartup(msg);

		if (msg.getState() == INSTALL) {
			installComponent(msg.getComponentId());
		} else {
			removeComponent(msg.getComponentId());
		}
	}

	private void installComponent(long componentId) {
		LOG.debug("Installing component: {} on entity: {}.", componentId, entityId);

		// Install the component.
		final ActorRef compActor = factory.startActor(getContext(), componentId);

		if (compActor == null) {
			LOG.warning("Component actor for comp id {} was not created.", componentId);
			return;
		}

		context().watch(compActor);
		componentActors.put(componentId, compActor);
	}

	private void removeComponent(long componentId) {
		LOG.debug("Removing component: {} on entity: {}.", componentId, entityId);

		// Remove the component.
		final ActorRef compRef = componentActors.getOrDefault(componentId, ActorRef.noSender());
		compRef.tell(PoisonPill.getInstance(), getSelf());
	}

	/**
	 * Check which actor was terminated and remove it.
	 * 
	 * @param term
	 *            The termination message from akka.
	 */
	private void handleTerminated(Terminated term) {
		componentActors.inverse().remove(term.actor());

		// Terminate if no components active anymore.
		if(componentActors.size() == 0) {
			getContext().stop(getSelf());
		}
	}

	/**
	 * Checks if we have already a defined entity id and if not we use it.
	 */
	private void checkStartup(EntityMessage msg) {
		if (entityId == 0) {
			entityId = msg.getEntityId();
		}
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/entity/EntityActor.kt
package net.bestia.zoneserver.actor.entity

import akka.actor.*
import com.google.common.collect.HashBiMap
import mu.KotlinLogging
import net.bestia.messages.entity.EntityMessage
import net.bestia.zoneserver.actor.Actor
import net.bestia.zoneserver.actor.AwaitResponseActor
import net.bestia.zoneserver.actor.entity.component.ComponentMessage
import net.bestia.zoneserver.actor.entity.component.ComponentRequest
import net.bestia.zoneserver.actor.entity.component.EntityComponentActorFactory
import net.bestia.zoneserver.actor.entity.component.UpdateComponent
import net.bestia.zoneserver.entity.Entity
import net.bestia.zoneserver.entity.component.Component

private val LOG = KotlinLogging.logger { }

/**
 * When send to the [EntityActor] it will update the given actor ref if a component
 * has changed.
 */
data class SubscribeForComponentUpdates(
    val componentType: Class<out Component>,
    val sendUpdateTo: ActorRef
)

/**
 * Used to notify all subscribers about a changed component.
 */
data class ComponentUpdated<T : Component>(
    val component: T
)

/**
 * Deletes a component from the Entity.
 */
data class DeleteComponent<T : Component>(
    override val entityId: Long,
    val componentClass: Class<T>
) : EntityMessage

data class NewEntity(
    val entity: Entity
) : EntityMessage {
  override val entityId: Long
    get() = entity.id
}

/**
 * Entity actor is stopped and all data also removed from database. This effectivly
 * deletes the entity.
 */
data class KillEntity(override val entityId: Long) : EntityMessage

/**
 * Entity actor is stopped but it is saved to the database before.
 */
data class SaveAndKillEntity(override val entityId: Long) : EntityMessage

/**
 * The [EntityActor] is a persistent actor managing all aspects of a
 * spawned entity. This means it will keep references to AI actors or attached
 * component actors. This actor has to be used as an persisted shared actor.
 *
 *
 * If there are no more active component actors this actor will cease operation.
 * However incoming component messages will restart this actor and attach the
 * component actor once more.
 *
 *
 * This actor has to react on certain incoming request messages like for example
 * attaching a ticking script to the entity.
 *
 * @author Thomas Felix
 */
@Actor
class EntityActor(
    private val factory: EntityComponentActorFactory
) : AbstractActor() {

  private val uninitialized: Receive = receiveBuilder()
      .match(NewEntity::class.java, this::setupEntity)
      .match(EntityRequest::class.java, this::handleUninitializedEntityRequest)
      .matchAny { context.stop(self) }
      .build()

  private val initialized: Receive = receiveBuilder()
      .match(NewEntity::class.java, this::setupEntity)
      .match(DeleteComponent::class.java, this::removeComponentActor)
      .match(UpdateComponent::class.java, this::updateComponentActor)
      .match(ComponentMessage::class.java, this::onComponentMessage)
      .match(SubscribeForComponentUpdates::class.java, this::subscribeForComponentUpdates)

      .match(Terminated::class.java, this::handleTerminatedComponentActor)
      .match(SaveAndKillEntity::class.java, this::handleSaveAndKill)
      .match(KillEntity::class.java) { context.stop(self) }
      .match(EntityRequest::class.java, this::handleExternalEntityRequest)
      .match(LocalEntityRequest::class.java, this::handleLocalEntityRequest)
      .build()

  private class ComponentActorCache {
    private val classToActor = HashBiMap.create<Class<Component>, ActorRef>()
    val activeComponentActorCount: Int get() = classToActor.size

    fun add(component: Component, compActor: ActorRef) {
      classToActor[component.javaClass] = compActor
    }

    fun allActors(): List<ActorRef> {
      return classToActor.values.toList()
    }

    fun remove(actor: ActorRef) {
      classToActor.inverse().remove(actor)
    }

    fun get(componentClass: Class<out Component>): ActorRef? {
      return classToActor[componentClass]
    }

    fun getAllCachedComponentClasses(): Set<Class<Component>> {
      return classToActor.keys.toSet()
    }
  }

  private val componentUpdateSubscriberCache = mutableSetOf<SubscribeForComponentUpdates>()
  private val componentActorCache = ComponentActorCache()
  private var entityId: Long = 0

  override fun createReceive(): Receive {
    return uninitialized
  }

  /**
   * We must cache any component subscription message as the actors are usually not re-sending it again
   * and the component actor might not be there already. We need to re-send it to the component actor if
   * an actor shows up.
   */
  private fun subscribeForComponentUpdates(msg: SubscribeForComponentUpdates) {
    componentUpdateSubscriberCache.add(msg)
  }

  private fun onComponentMessage(msg: ComponentMessage<*>) {
    val componentActor = componentActorCache.get(msg.componentType)

    if (componentActor == null) {
      LOG.debug { "Component missing: ${msg.componentType}" }
      return
    }

    componentActor.tell(msg, self)
  }

  private fun setupEntity(msg: NewEntity) {
    LOG.trace { "Creating entity actor: $msg" }
    entityId = msg.entity.id
    msg.entity.allComponents.forEach { createComponentActor(it) }

    context.become(initialized)
  }

  private fun handleUninitializedEntityRequest(msg: EntityRequest) {
    msg.replyTo.tell(EntityDoesNotExist, self)
    context.stop(self)
  }

  private fun handleLocalEntityRequest(msg: LocalEntityRequest) {
    handleEntityRequest(msg.replyTo, msg.context)
  }

  private fun handleExternalEntityRequest(msg: EntityRequest) {
    handleEntityRequest(msg.replyTo, msg.context)
  }

  private fun handleEntityRequest(requester: ActorRef, requestCtx: Any?) {
    val awaitedComponentClasses = componentActorCache.getAllCachedComponentClasses()

    val hasAllResponses = { receivedResponses: List<Any> ->
      receivedResponses.map { it.javaClass }
          .toSet() == awaitedComponentClasses
    }

    val waitResponseProps = AwaitResponseActor.props(hasAllResponses) {
      @Suppress("UNCHECKED_CAST")
      val entity = Entity.withComponents(
          entityId,
          it.getAllResponses()
      )
      val entityResponse = EntityResponse(
          entity,
          requestCtx
      )
      requester.tell(entityResponse, self)
    }
    val waitResponseActor = context.actorOf(waitResponseProps)

    componentActorCache.allActors().forEach {
      val requestMsg = ComponentRequest(replyTo = waitResponseActor)
      it.tell(requestMsg, self)
    }
  }

  private fun updateComponentActor(msg: UpdateComponent<*>) {
    val componentActor = componentActorCache.get(msg.component.javaClass)
        ?: createComponentActor(msg.component)
        ?: return

    LOG.trace { "Updating component: $msg on entity: $entityId" }
    componentActor.tell(msg.component, sender())
  }

  private fun removeComponentActor(msg: DeleteComponent<*>) {
    LOG.trace { "Removing component: ${msg.componentClass.simpleName} on entity: $entityId" }
    componentActorCache.get(msg.componentClass)?.tell(PoisonPill.getInstance(), self)
  }

  private fun createComponentActor(component: Component): ActorRef? {
    val existingCompActor = componentActorCache.get(component.javaClass)

    if (existingCompActor != null) {
      return existingCompActor
    }

    return factory.startActor(context, component)?.also {
      context().watch(it)
      componentActorCache.add(component, it)
      componentUpdateSubscriberCache.forEach { subMsg -> it.tell(subMsg, self) }
    }
  }

  /**
   * Terminate itself when there is no active component anymore.
   */
  private fun handleTerminatedComponentActor(term: Terminated) {
    componentActorCache.remove(term.actor)
    if (componentActorCache.activeComponentActorCount == 0) {
      context.stop(self)
    }
  }

  private fun handleSaveAndKill(msg: SaveAndKillEntity) {
    // TODO Persist the given entity.
    componentActorCache.allActors().forEach {
      // Make own message
      it.tell(msg, sender)
    }

    context.stop(self)
  }
}
